homework-submission/homework-week-2/react.js

```diff
@@ -0,0 +1,125 @@
+/* eslint-disable no-undef */
+/* eslint-disable react/react-in-jsx-scope */
+/* eslint-disable react/jsx-no-undef */
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.addTodo = this.addTodo.bind(this);
+  }
+
+  state = {
+    todos: [
+      {
+        id: 1,
+        description: 'Get out of bed',
+        deadline: '2017-09-11',
+        done: true
+      },
+      {
+        id: 2,
+        description: 'Brush teeth',
+        deadline: '2017-09-10',
+        done: false
+      },
+      {
+        id: 3,
+        description: 'Eat breakfast',
+        deadline: '2017-09-09',
+        done: false
+      }
+    ]
+  };
+
+  addTodo(e) {
+    if (!this.desc.value || !this.date.value) {
+      alert('Each field must be filled.\nPlease check your input !');
+      e.preventDefault();
+    } else {
+      const todos = this.state.todos;
+      const description = this.desc.value;
+      const deadline = this.date.value;
+      const lastTodoId = todos[todos.length - 1].id;
+      const id = lastTodoId + 1;
+      const newTodo = {
+        id: id,
+        description: description,
+        deadline: deadline,
+        done: false
+      };
+      todos.push(newTodo);
+      this.setState({ todos });
+      e.preventDefault();
+    }
+  }
+
+  doneUndone = index => {
+    const { todos } = this.state;
+
+    const newTodos = todos.map((todo, i) => {
+      if (index === i) {
+        todo.done = !todo.done;
+      }
+      return todo;
+    });
+    this.setState({ todos: newTodos });
+  };
+
+  deleteTodo = index => {
+    const { todos } = this.state;
+    delete todos[index];
```

Comment:
nooooooo, don't use delete, it mutates the array, use concat, or splice to not mutate the original array.

Reply:
Replaced by splice but, I can't understand the logic. Why it must be non-mutating ? and splice method is a mutating method as well.

Reply:
Replaced by splice method but, I can't get the logic. Why it must be non-mutating ? Splice method is a mutating one as well.